src/Interpreters/TautologicalComparisonOptimizer.cpp

```diff
@@ -0,0 +1,48 @@
+#include "TautologicalComparisonOptimizer.h"
+
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTLiteral.h>
+
+#include <Functions/FunctionsComparison.h>
+
+namespace DB
+{
+
+bool TautologicalComparisonOptimizer::needChildVisit(DB::ASTPtr &, const DB::ASTPtr &)
+{
+    return true;
+}
+
+void TautologicalComparisonOptimizer::visit(ASTPtr & ast, Data &)
+{
+    static std::unordered_set<std::string_view> comparison_functions
+    {
+        NameEquals::name,
+        NameNotEquals::name,
+        NameLess::name,
+        NameGreater::name,
+        NameLessOrEquals::name,
+        NameGreaterOrEquals::name
+    };
+
+    if (const auto * func = ast->as<ASTFunction>())
+    {
+        auto it = comparison_functions.find(func->name);
+        if (it == comparison_functions.end())
+            return;
+
+        auto & lhs = func->arguments->children[0];
+        auto & rhs = func->arguments->children[1];
+
+        auto lhs_hash = lhs->getTreeHash();
+        auto rhs_hash = rhs->getTreeHash();
+
+        if (likely(lhs_hash != rhs_hash))
+            return;
+
+        func->arguments->children[0] = std::make_shared<ASTLiteral>(1);
+        func->arguments->children[1] = std::make_shared<ASTLiteral>(1);
```

Comment:
Ok, but maybe we can replace the whole function to literal 1 or literal 0

Reply:
@KochetovNicolai right now seems easier to understand what is going on.